pkg/repository/v1_repository.go

```diff
@@ -50,32 +60,134 @@ func NewV1Repo(mirror Mirror, opts Options, local v1manifest.LocalManifests) (*V
 		local:   local,
 	}
 
-	if err := repo.loadRoot(); err != nil {
-		return nil, errors.AddStack(err)
-	}
-
-	return repo, nil
+	return repo
 }
 
 const maxTimeStampSize uint = 1024
 const maxRootSize uint = 1024 * 1024
 
-func (r *V1Repository) loadRoot() error {
-	root := new(v1manifest.Root)
-	exists, err := r.local.LoadManifest(root)
+// UpdateComponents updates the components described by specs.
+func (r *V1Repository) UpdateComponents(specs []ComponentSpec) error {
+	_, err := r.ensureManifests()
 	if err != nil {
-		return errors.AddStack(err)
+		return err
 	}
+	for _, spec := range specs {
+		manifest, err := r.updateComponentManifest(spec.ID)
+		if err != nil {
+			// TODO could report error and continue
+			return err
+		}
+		if manifest == nil {
+			continue
+		}
 
-	if !exists {
-		return errors.New("no trusted root in the local manifest")
+		platform := utils.PlatformString()
+		versions, ok := manifest.Platforms[platform]
+		if !ok {
+			// TODO could report error and continue
+			return fmt.Errorf("platform %s not supported by component %s", platform, spec.ID)
+		}
+
+		version, versionItem, err := r.selectVersion(spec.ID, versions, spec.Version)
+		if err != nil {
+			// TODO could report error and continue
+			return err
+		}
+
+		if !spec.Force {
+			installed, err := r.local.ComponentInstalled(spec.ID, version)
+			if err != nil {
+				return err
+			}
+			if installed {
+				// TODO could report error and continue
+				return fmt.Errorf("component %s version %s is already installed", spec.ID, version)
+			}
+		}
+
+		reader, err := r.downloadComponent(versionItem)
```

Comment:
defer reader.Close()

Reply:
This is just a Reader, not a ReadCloser and doesn't need (and can't be) closed